### ISSUE TITLE
Out of space error while installing npm packages

### ISSUE THREAD

**Issue by hashemi** [progrium/dokku#1673]
I'm getting an out of space error trying to push an Ember project with a custom Dockerfile to my dokku (1gb droplet on DO). It works fine on my Mac, when I do `docker build .`.

I've done diagnostics on the host droplet and everything seems to be fine. Googling the problem turns up people who actually ran out of space or inodes but that doesn't seem to be the problem here.

```
Log during deployment
npm ERR! tar.unpack untar error /root/.npm/ember-cli/1.13.8/package.tgz
npm ERR! Linux 3.13.0-57-generic
npm ERR! argv "/usr/local/bin/node" "/usr/local/bin/npm" "install" "-g" "ember-cli" "bower"
npm ERR! node v4.2.2
npm ERR! npm  v2.14.7
npm ERR! path /usr/local/lib/node_modules/ember-cli/node_modules/testem/node_modules/tap-parser/test/fixtures/skip-one-fail.tap
npm ERR! code ENOSPC
npm ERR! errno -28
npm ERR! syscall open

npm ERR! nospc ENOSPC: no space left on device, open '/usr/local/lib/node_modules/ember-cli/node_modules/testem/node_modules/tap-parser/test/fixtures/skip-one-fail.tap'
npm ERR! nospc This is most likely not a problem with npm itself
npm ERR! nospc and is related to insufficient space on your system.

Dockerfile
# nginx
FROM debian:jessie

RUN apt-key adv --keyserver hkp://pgp.mit.edu:80 --recv-keys 573BFD6B3D8FBC641079A6ABABF5BD827BD9BF62
RUN echo "deb http://nginx.org/packages/mainline/debian/ jessie nginx" >> /etc/apt/sources.list

ENV NGINX_VERSION 1.9.6-1~jessie

RUN apt-get update && \
    apt-get install -y ca-certificates nginx=${NGINX_VERSION} && \
    rm -rf /var/lib/apt/lists/*

# forward request and error logs to docker log collector
RUN ln -sf /dev/stdout /var/log/nginx/access.log
RUN ln -sf /dev/stderr /var/log/nginx/error.log

VOLUME ["/var/cache/nginx"]

EXPOSE 80 443

CMD ["nginx", "-g", "daemon off;"]

# jessie-curl
RUN apt-get update && apt-get install -y --no-install-recommends \
    ca-certificates \
    curl \
    wget \
  && rm -rf /var/lib/apt/lists/*

# node

# gpg keys listed at https://github.com/nodejs/node
RUN set -ex \
  && for key in \
    9554F04D7259F04124DE6B476D5A82AC7E37093B \
    94AE36675C464D64BAFA68DD7434390BDBE9B9C5 \
    0034A06D9D9B0064CE8ADF6BF1747F4AD2306D93 \
    FD3A5288F042B6850C66B31F09FE44734EB7990E \
    71DCFD284A79C3B38668286BC97EC7A07EDE3FC1 \
    DD8F2338BAE7501E3DD5AC78C273792F7D83545D \
  ; do \
    gpg --keyserver ha.pool.sks-keyservers.net --recv-keys "$key"; \
  done

ENV NPM_CONFIG_LOGLEVEL info
ENV NODE_VERSION 4.2.2

RUN curl -SLO "https://nodejs.org/dist/v$NODE_VERSION/node-v$NODE_VERSION-linux-x64.tar.gz" \
  && curl -SLO "https://nodejs.org/dist/v$NODE_VERSION/SHASUMS256.txt.asc" \
  && gpg --verify SHASUMS256.txt.asc \
  && grep " node-v$NODE_VERSION-linux-x64.tar.gz\$" SHASUMS256.txt.asc | sha256sum -c - \
  && tar -xzf "node-v$NODE_VERSION-linux-x64.tar.gz" -C /usr/local --strip-components=1 \
  && rm "node-v$NODE_VERSION-linux-x64.tar.gz" SHASUMS256.txt.asc


# The ember app

# toolchain required to build one of ember's dependencies using node-gyp
RUN apt-get update && apt-get install -y --no-install-recommends \
    python \
    build-essential \
  && rm -rf /var/lib/apt/lists/*

RUN npm install -g ember-cli bower

COPY ember.conf /etc/nginx/conf.d/ember.conf

RUN mkdir /app
COPY . /app

RUN cd /app \
  && npm install \
  && bower --allow-root install \
  && ember build --environment="production" \
  && rm -rf /usr/share/nginx/html \
  && mv dist /usr/share/nginx/html

RUN apt-get remove --purge -y \
    python \
    build-essential \
  && apt-get -y autoremove \
  && apt-get -y clean \
  && rm -rf /app

root# df
Filesystem     1K-blocks    Used Available Use% Mounted on
/dev/vda1       30830588 5649544  23591908  20% /
none                   4       0         4   0% /sys/fs/cgroup
udev              498080       4    498076   1% /dev
tmpfs             101788     356    101432   1% /run
none                5120       0      5120   0% /run/lock
none              508936    1960    506976   1% /run/shm
none              102400       0    102400   0% /run/user

root# df -i
Filesystem      Inodes  IUsed   IFree IUse% Mounted on
/dev/vda1      1966080 246825 1719255   13% /
none            127234     11  127223    1% /sys/fs/cgroup
udev            124520    402  124118    1% /dev
tmpfs           127234    329  126905    1% /run
none            127234      1  127233    1% /run/lock
none            127234     10  127224    1% /run/shm
none            127234      2  127232    1% /run/user

root# uname -a
Linux HOSTNAME 3.13.0-57-generic #95-Ubuntu SMP Fri Jun 19 09:28:15 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux

root# docker version
Client:
 Version:      1.9.0
 API version:  1.21
 Go version:   go1.4.2
 Git commit:   76d6bc9
 Built:        Tue Nov  3 17:43:42 UTC 2015
 OS/Arch:      linux/amd64

Server:
 Version:      1.9.0
 API version:  1.21
 Go version:   go1.4.2
 Git commit:   76d6bc9
 Built:        Tue Nov  3 17:43:42 UTC 2015
 OS/Arch:      linux/amd64

root# docker -D info
Containers: 4
Images: 63
Server Version: 1.9.0
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 71
 Dirperm1 Supported: false
Execution Driver: native-0.2
Logging Driver: json-file
Kernel Version: 3.13.0-57-generic
Operating System: Ubuntu 14.04.3 LTS
CPUs: 1
Total Memory: 994 MiB
Name: HOSTNAME
ID: JGKG:C4PX:RQRD:BMJV:QTH2:FPM5:IE3D:SHIW:2RU2:WGWE:72AX:2J2U
WARNING: No swap limit support

root# dokku version
0.4.3

root# dokku plugin
  00_dokku-standard    0.4.0 enabled    dokku core standard plugin
  20_events            0.4.0 enabled    dokku core events logging plugin
  apps                 0.4.0 enabled    dokku core apps plugin
  backup               0.4.0 enabled    dokku core backup plugin
  build-env            0.4.0 enabled    dokku core build-env plugin
  certs                0.4.0 enabled    dokku core certificate management plugin
  checks               0.4.0 enabled    dokku core checks plugin
  common               0.4.0 enabled    dokku core common plugin
  config               0.4.0 enabled    dokku core config plugin
  docker-options       0.4.0 enabled    dokku core docker-options plugin
  domains              0.4.0 enabled    dokku core domains plugin
  enter                0.4.0 enabled    dokku core enter plugin
  git                  0.4.0 enabled    dokku core git plugin
  logs                 0.4.0 enabled    dokku core logs plugin
  mariadb              1.0.0 enabled    dokku mariadb service plugin
  named-containers     0.4.0 enabled    dokku core named containers plugin
  nginx-vhosts         0.4.0 enabled    dokku core nginx-vhosts plugin
  plugin               0.4.0 enabled    dokku core plugin plugin
  ps                   0.4.0 enabled    dokku core ps plugin
  shell                0.4.0 enabled    dokku core shell plugin
  tags                 0.4.0 enabled    dokku core tags plugin
  tar                  0.4.0 enabled    dokku core tar plugin
```

**Comment by josegonzalez** [comment]
Can you provide a sample app that has this issue?

**Comment by hashemi** [comment]
It's my own app which I just started working on. Right now it's a stock Ember app with bootstrap as a bower component and one route. I will try to reproduce the error with a stock Ember app.

**Comment by hashemi** [comment]
After some experimentation, I was able to replicate the problem with this Dockerfile:

``` Dockerfile
FROM node:4.2.2-slim
RUN npm install -g ember-cli bower
```

It may have been a coincidence, but the problem went away after stopping an unrelated PHP/mariadb container. However, I started running into insufficient memory issues. It eventually worked, after several retries. I watched as NPM consume 98% of my VPSs CPU and memory during the process. The solution thrown around to the memory problem is to enable a swap partition. Instead, I'm going to forego server side npm for now.

I'm closing this as I don't believe it's a dokku issue.
